dvc/dvcfile.py

```diff
@@ -332,6 +332,24 @@ def latest_version_info(self):
         version = LOCKFILE_VERSION.V2.value  # pylint:disable=no-member
         return {SCHEMA_KWD: version}
 
+    def _remove_unused_outs_in_gitignore(
+        self, orig_stage: dict, new_stage: dict
+    ):
+        def get_out_path_set(stage_data: dict) -> set:
+            stage_outs = stage_data.get("outs", [])
```

Comment:
The issue #5156 is pretty broad, though I like your approach of using `dvc.lock` data to figure out what to remove.

Though I see a few problems regarding the implementation:
1. Lockfiles don't have `wdir` (it's read from `dvc.yaml`). So we don't have a way to know where the `.gitignore` file is. You cannot reliably use `wdir` from `dvc.yaml` or `stage` arg passed in `lockfile.dump()` as they might have been updated/changed/removed as well. We might end up deleting the user's `.gitignore` entries this way even if we try.
2. `Lockfile` does not feel like a good place to determine this (see that you have to parse `outs`). We need some abstraction to hide this somewhere. 

 (Maybe we need a synchronized dumping abstraction that takes care of this stuff together. We have been discussing of introducing a `repo.index.update()` abstraction that might take the responsibility, but we are still not sure of it.)

Reply:
Please ignore the 2) point, just trying to provide the context. We can accept the contribution as-is for now, if we are able to solve 1) somehow.

Reply:
@skshetry 
I agreed, file operations ( `.gitignore`, `dvc.yaml`, `dvc.lock`) are too scattered and hard to maintain. It took me half an hour to found where the outputs are added to `.gitignore`.

Reply:
The same with #5293. We need to summarize the DVC index files (`dvc.yaml`, `dvc.lock`) information to make the final modification decision.

Reply:
@karajan1001, we don't have enough information anyway. We cannot depend on `dvc.yaml` data for this at all, as this is the "source of truth" that resulted in old .gitignore entries being no longer valid. So we cannot trust it, not even for `wdir` as it could have changed as well.

Feels like keeping 3 kinds of data format (dvc.lock, dvc.yaml and .gitignore) in sync is not a good idea?

Reply:
@skshetry 
Not only `.gitignore` files but also `dvc.yaml` and `dvc.lock` can we have inside a `wdir`. Data tracked in one `dvc.yaml` might be moved to another one. Gathering all of the information from these dvcfiles and lockfiles and keep them in sync is not an easy work.